File: unijava/src/org/unified/uni/lexer/LexicalToken.java
package org.unified.uni.lexer;

import com.sun.istack.internal.NotNull;

import java.util.ArrayList;

/**
 * Uni Lexical Token
 * <p>
 * Created by Michael Vlasov on 19.04.2018.
 * <p>
 * Copyright (c) 2018 MichaelVlasov
 */
public class LexicalToken {
	public final int name;


	public String nameString() {
		return names.get(name);
	}


	@NotNull
	public String valueString() {
		return value != null ? value.toString() : "";
	}


	public int valueInt() {
		Integer intValue = (Integer) value;
		return intValue != null ? intValue : 0;
	}


	public LexicalToken with(Object value) {
		return new LexicalToken(name, value);
	}


	public static LexicalToken register(String nameString) {
		int name = LexicalToken.names.size();
		LexicalToken.names.add(nameString);
		return new LexicalToken(name, null);
	}


	@Override
	public String toString() {
		String name = LexicalToken.names.get(this.name);
		return value != null ? "(" + nameString() + ", " + value.toString() + ")" : name;
	}


	// Internals

	private Object value;


	private LexicalToken(int name, Object value) {
		this.name = name;
		this.value = value;
	}


	private static ArrayList<String> names = new ArrayList<>();
}

File: uniswift/Uni/Uni/Lexer/LexicalToken.swift
//
//  Token.swift
//  Uni
//
//  Created by Michael Vlasov on 28.04.2018.
//  Copyright © 2018 Unified. All rights reserved.
//

import Foundation

public struct LexicalToken: CustomStringConvertible {
	public enum Value {
		case missing
		case string(String)
		case int(Int)
	}
	public let name: Int
	public let value: Value
	
	private init(name: Int, value: Value) {
		self.name = name
		self.value = value
	}
	
	public var nameString: String {
		return LexicalToken.names[name]
	}
	
	public var valueString: String {
		switch value {
		case .missing:
			return ""
		case .string(let string):
			return string
		case .int(let int):
			return String(describing: int)
		}
	}
	
	public var valueInt: Int {
		switch value {
		case .missing:
			return 0
		case .string(let string):
			return Int(string) ?? 0
		case .int(let int):
			return int
		}
	}
	
	public func with(value: Value) -> LexicalToken {
		return LexicalToken(name: name, value: value)
	}
	
	
	public var description: String {
		if case .missing = value {
			return nameString
		}
		return "(\(nameString), \(valueString))"
	}
	
	// MARK: - Static

	public static func register(_ nameString: String) -> LexicalToken {
		let name = LexicalToken.names.count
		LexicalToken.names.append(nameString)
		return LexicalToken(name: name, value: .missing)
	}
	
	
	public static let empty = LexicalToken(name: 0, value: .missing)
	
	// MARK: - Internals
	
	private static var names = [""]
}


